warehouse/utils/security_policy.py

```diff
@@ -0,0 +1,95 @@
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from pyramid.authorization import Authenticated, Everyone
+from pyramid.interfaces import ISecurityPolicy
+from zope.interface import implementer
+
+from warehouse.accounts.interfaces import IUserService
+
+
+class SecurityPolicy:
+    def __init__(self, callback=None):
+        self._callback = callback
+
+    def identity(self, request):
+        login_service = request.find_service(IUserService, context=None)
+        user = login_service.get_user(self.unauthenticated_userid(request))
+        if user is None:
+            return None
+
+        principals = []
+        if self._callback is not None:
+            principals = self._callback(user.id, request)
+            if principals is None:
+                return None
+
+        return {"entity": user, "principals": principals}
```

Comment:
I think I'd prefer this to just return the ORM object that corresponds to the identity (here, the user). The computation of the principals can happen when we need it in `permits` instead, we don't need to pass it around, and we don't need principals available everywhere `request.identity` would be available.

I also think the `identity` function for `SecurityPolicy` should probably be `NotImplemented` and the subclass should be responsible for generating the corresponding identity instead. This means that `SessionSecurityPolicy` and `BasicAuthSecurityPolicy` could always return `Users`, but we could eventually update `MacaroonSecurityPolicy` to return other identities.

Since for now, they would all share the same logic for computing a user from a request, perhaps we turn this into a standalone function or mixin or something instead.

Reply:
Yeah, this is another case where I cargo-culted it from Pyramid's porting guide: they recommended having the identity contain the principles, but I agree that there's no reason for it to. I'll refactor so that it's just the ORM model.

Reply:
Oh, I remember why I did this now -- our AuthZ policies are still using the old APIs, which require the principals to be passed in explicitly. I couldn't figure out how else to get them into those APIs without stashing them on the `identity`, since the only context available is `request`, `context` (already set), and `permissions` (already set?)

Reply:
I think they can be computed in `permits` right before we pass them to the AuthZ policy, no?

Reply:
Yeah, I think you're right (I was overthinking it). `permits` can use `request.identity` as long as it's present, so we can do anything that the policy's `identity` would do anyways. I'll refactor!

Reply:
[0b788d9](https://github.com/pypa/warehouse/pull/11218/commits/0b788d91af438990ebd4961eb070c0ce31af468e)